### ISSUE TITLE
Accessibility issue with filter criteria in ReactiveList 

### ISSUE THREAD

**Issue by allendol** [appbaseio/reactivesearch#1129]
We are using ReactiveList. The filter criteria is not accessible using the key tab.
When using the key tab, the filter criteria is skipped. No way to get focus on it using the keyboard.
![FilterCriteria](https://user-images.githubusercontent.com/40170767/63964450-2010f280-ca65-11e9-8fe1-729c95389d6b.png)

**Comment by jyash97** [comment]
Thanks for reporting issues on accessibility will take a look into this and create a fix for this.

**Comment by allendol** [comment]
@jyash97,
Below is my selectedfilters.js file. It is much different than the one is written in #1132
---------------------

'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _emotionTheming = require('emotion-theming');

var _actions = require('@appbaseio/reactivecore/lib/actions');

var _types = require('@appbaseio/reactivecore/lib/utils/types');

var _types2 = _interopRequireDefault(_types);

var _helper = require('@appbaseio/reactivecore/lib/utils/helper');

var _Button = require('../../styles/Button');

var _Button2 = _interopRequireDefault(_Button);

var _Container = require('../../styles/Container');

var _Container2 = _interopRequireDefault(_Container);

var _Title = require('../../styles/Title');

var _Title2 = _interopRequireDefault(_Title);

var _utils = require('../../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SelectedFilters = function (_Component) {
	_inherits(SelectedFilters, _Component);

	function SelectedFilters() {
		var _temp, _this, _ret;

		_classCallCheck(this, SelectedFilters);

		for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
			args[_key] = arguments[_key];
		}

		return _ret = (_temp = (_this = _possibleConstructorReturn(this, _Component.call.apply(_Component, [this].concat(args))), _this), _this.componentDidUpdate = function () {
			if (_this.props.onChange) {
				_this.props.onChange(_this.props.selectedValues);
			}
		}, _this.remove = function (component) {
			var value = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
			var onClear = _this.props.onClear;

			_this.props.setValue(component, null);
			if (onClear) {
				onClear(component, value);
			}
		}, _this.clearValues = function () {
			var onClear = _this.props.onClear;

			_this.props.clearValues();
			if (onClear) {
				onClear(null);
			}
		}, _this.renderValue = function (value, isArray) {
			if (isArray && value.length) {
				var arrayToRender = value.map(function (item) {
					return _this.renderValue(item);
				});
				return arrayToRender.join(', ');
			} else if (value && (typeof value === 'undefined' ? 'undefined' : _typeof(value)) === 'object') {
				// TODO: support for NestedList
				var label = (typeof value.label === 'string' ? value.label : value.value) || value.key || value.distance || null;
				if (value.location) {
					label = value.location + ' - ' + label;
				}
				return label;
			}
			return value;
		}, _this.renderFilters = function () {
			var selectedValues = _this.props.selectedValues;

			return Object.keys(selectedValues).filter(function (id) {
				return _this.props.components.includes(id) && selectedValues[id].showFilter;
			}).map(function (component, index) {
				var _selectedValues$compo = selectedValues[component],
				    label = _selectedValues$compo.label,
				    value = _selectedValues$compo.value,
				    category = _selectedValues$compo.category;

				var isArray = Array.isArray(value);

				if (label && (isArray && value.length || !isArray && value)) {
					var valueToRender = category ? _this.renderValue(value + ' in ' + category + ' category', isArray) : _this.renderValue(value, isArray);
					return _react2.default.createElement(
						_Button2.default,
						{
							className: (0, _helper.getClassName)(_this.props.innerClass, 'button') || null,
							key: component + '-' + (index + 1),
							onClick: function onClick() {
								return _this.remove(component, value);
							}
						},
						_react2.default.createElement(
							'span',
							null,
							selectedValues[component].label,
							': ',
							valueToRender
						),
						_react2.default.createElement(
							'span',
							null,
							'\u2715'
						)
					);
				}
				return null;
			}).filter(Boolean);
		}, _temp), _possibleConstructorReturn(_this, _ret);
	}

	SelectedFilters.prototype.render = function render() {
		if (this.props.render) {
			return this.props.render(this.props);
		}

		var theme = this.props.theme;

		var filtersToRender = this.renderFilters();
		var hasValues = !!filtersToRender.length;

		return _react2.default.createElement(
			_Container2.default,
			{
				style: this.props.style,
				className: (0, _Button.filters)(theme) + '\n\t\t\t\t' + (this.props.className || '')
			},
			this.props.title && hasValues && _react2.default.createElement(
				_Title2.default,
				{ className: (0, _helper.getClassName)(this.props.innerClass, 'title') || null },
				this.props.title
			),
			filtersToRender,
			this.props.showClearAll && hasValues ? _react2.default.createElement(
				_Button2.default,
				{
					className: (0, _helper.getClassName)(this.props.innerClass, 'button') || null,
					onClick: this.clearValues
				},
				this.props.clearAllLabel
			) : null
		);
	};

	return SelectedFilters;
}(_react.Component);

SelectedFilters.propTypes = {
	clearValues: _types2.default.func,
	setValue: _types2.default.func,
	components: _types2.default.components,
	selectedValues: _types2.default.selectedValues,
	className: _types2.default.string,
	clearAllLabel: _types2.default.title,
	innerClass: _types2.default.style,
	showClearAll: _types2.default.bool,
	style: _types2.default.style,
	theme: _types2.default.style,
	onClear: _types2.default.func,
	render: _types2.default.func,
	title: _types2.default.title,
	onChange: _types2.default.func
};

SelectedFilters.defaultProps = {
	className: null,
	clearAllLabel: 'Clear All',
	showClearAll: true,
	style: {}
};

var mapStateToProps = function mapStateToProps(state) {
	return {
		components: state.components,
		selectedValues: state.selectedValues
	};
};

var mapDispatchtoProps = function mapDispatchtoProps(dispatch) {
	return {
		clearValues: function clearValues() {
			return dispatch((0, _actions.clearValues)());
		},
		setValue: function setValue(component, value) {
			return dispatch((0, _actions.setValue)(component, value));
		}
	};
};

var ConnectedComponent = (0, _utils.connect)(mapStateToProps, mapDispatchtoProps)((0, _emotionTheming.withTheme)(function (props) {
	return _react2.default.createElement(SelectedFilters, _extends({ ref: props.myForwardedRef }, props));
}));

// eslint-disable-next-line
exports.default = _react2.default.forwardRef(function (props, ref) {
	return _react2.default.createElement(ConnectedComponent, _extends({}, props, { myForwardedRef: ref }));
});

**Comment by jyash97** [comment]
Can you please specify what is different and what did you change as this isnt helpful as you copied  all the code directly from node_modules

**Comment by allendol** [comment]
@jyash97,
I didn't change anything. I was trying to say the source code of selectedfilters.js file that I have is different than the one updated in #1132
I just copied and pasted the whole source code of my selectedfilters.js file in above which is not changed.
I think if I want to correspond this with the code in #1132 , may be the following part should be updated on my selectedfilters.js file version:
like 107:
						_Button2.default,
						{
							className: (0, _helper.getClassName)(_this.props.innerClass, 'button') || null,
							key: component + '-' + (index + 1),
							onClick: function onClick() {
								return _this.remove(component, value);
							}
						},

and the line 156 which is:
			this.props.showClearAll && hasValues ? _react2.default.createElement(
				_Button2.default,
				{
					className: (0, _helper.getClassName)(this.props.innerClass, 'button') || null,
					onClick: this.clearValues
				},
				this.props.clearAllLabel
			) : null

Can you just tell me on like 110 how I can add:
							tabIndex="0"
							onKeyPress={event => handleA11yAction(event, () => this.remove(component, value))}
The above two lines are extracted from your solution in 1132
and also on line 160: how I can add below:
						tabIndex="0"
						onKeyPress={event => handleA11yAction(event, this.clearValues)}
Again the above two lines are form your solution in 1132.

**Comment by jyash97** [comment]
Any specific reason you want to change things on `node_modules` or are you just want to test the changes done in PR #1132 ?

**Comment by allendol** [comment]
@jyash97 ,
- In my original bug report I wrote there is accessibility issue in filter criteria in ReactiveSearch.
- I believe you guys find a way to fix it and write in #1132 
- It is my understanding that in #1132 you have located where the changes has to apply.
- I think the changes should be applied in selectedfilters.js
- So I compared my selectedfilters.js with selectedfilters.js defined in #1132 and saw there is difference.
- I located the lines that has to be changed in my version of selectedfilters.js to fix the accessiblity problem.
- My request was how to add couple of lines in my version of selectedfilters.js to fix accessibility issue.

**Comment by allendol** [comment]
@jyash97 
Any feedback? I still don't know how I can change my selectedfilters.js.

**Comment by allendol** [comment]
@bietkul,
I tried to add this:
tabIndex="0"
onKeyPress={event => handleA11yAction(event, () => this.remove(component, 
value))}

But below is the format of my selectedfilters.js:
----
						_Button2.default,
						{
							className: (0, _helper.getClassName)(_this.props.innerClass, 'button') || null,
							key: component + '-' + (index + 1),
							onClick: function onClick() {
								return _this.remove(component, value);
							}
						},
-----
The format is different than your selectedfilters.js. Your selectedfilters.js looks like below:
-----
						<Button							<Button
							className={getClassName(this.props.innerClass, 'button') || null}								className={getClassName(this.props.innerClass, 'button') || null}
							key={`${component}-${index + 1}`}								key={`${component}-${index + 1}`}
							tabIndex="0"
							onKeyPress={event => handleA11yAction(event, () => this.remove(component, value))}
							onClick={() => this.remove(component, value)}								onClick={() => this.remove(component, value)}
						>							>
------
So I need to translate / convert:
tabIndex="0"
onKeyPress={event => handleA11yAction(event, () => this.remove(component, 

To the format that works in my selectedfilters.js.
How can I do it?

**Comment by bietkul** [comment]
@jyash97 Can you check on it?

**Comment by jyash97** [comment]
Hey @allendol 
This changes would be added in the next release you can then just upgrade your version and run the app. Making changes to `node_modules` is not good way to go about. But if you still want to make the changes in `node_modules` you can setup this repo locally and copy the files from respective library files.

**Comment by allendol** [comment]
Hello @jyash97 
I am at deployment phase and I am afraid if I upgrade the version, it would cause something else to break.
All I want is how I can modify these two lines:
tabIndex="0"
onKeyPress={event => handleA11yAction(event, () => this.remove(component, value))}
to work at:
_Button2.default,
    {
        className: (0, _helper.getClassName)(_this.props.innerClass, 'button') || null,
        key: component + '-' + (index + 1),
        onClick: function onClick() {
            return _this.remove(component, value);
        }
    },

Most probably 
tabIndex="0" 
should be changed to 
tabIndex:"0",
But how about:
onKeyPress={event => handleA11yAction(event, () => this.remove(component, value))}

**Comment by jyash97** [comment]
Can you try with this code: https://gist.github.com/jyash97/6b12e343ebc68a0e298b49b8d232fb13

If this doesnt work. You can follow these steps:

1. Setup Project Locally. Check this: https://github.com/appbaseio/reactivesearch/blob/next/.github/CONTRIBUTING.md

2. Run `yarn dev:web`

3. Go to `packages/web/lib/components/basic/SelectedFilters` and check the code that has changed.

I highly recommend upgrading the version you can always test if things are breaking but if you are already on `rc` version it is less likely that things break after updating to version 3 or higher.
